### ISSUE TITLE
intermittent 502 (possible TCP timeout issue?)

### ISSUE THREAD

**Issue by nquandt** [microsoft/reverse-proxy#1955]
### Describe the bug
Azure App Gateway -> Kubernetes hosted[ YARP (As Cluster Reverse-Proxy + Ingress) -> Other kube services ]
![image](https://user-images.githubusercontent.com/57432850/206313674-b5803608-e500-4243-b340-03ad2f5dc12e.png)

Here is a general overview of the infra, I just removed some names of actual services

I am hosting YARP on a kubernetes cluster as both a reverse proxy and ingress controller. It is behind an Azure App Gateway (not my decision) that essentially is just a firewall step. So all requests to the App Gateway get immediately proxied to YARP and Yarp handles routing across our cluster. 

Occasionally I experience "random" 502's from AAG telling me my backend pool is doing something wrong. But a quick refresh and it will resolve. After doing some reading I am wondering if I am experiencing something similar to [this](https://iximiuz.com/en/posts/reverse-proxy-http-keep-alive-and-502s/). 

### To Reproduce
Setup AAG with backend pool of kubernetes( on aks ) YARP as exposed ingress (virtual net) with another app on kubernetes to route to.

Make some requests.

Eventually some will randomly give 502

### Possible idea
I don't think this is technically a fix, but I considered reducing YARP's keep alive timeouts so that it closes connections before AAG. I'm not sure if its a placebo, but doing this has seemed to reduce the number of occurrences of 502.

```csharp
builder.WebHost.UseKestrel(options => 
{
	options.Limits.KeepAliveTimeout = TimeSpan.FromSeconds(5);
});
```

### Further technical details

- using code from main branch Yarp.ReverseProxy 1.1.1
- docker image mcr.microsoft.com/dotnet/aspnet:7.0
- I get alot of these messages in my logs (on YARP) telling me that AAG is closing connections
```bash
[22:34:47 DBG] Connection id "0HMMOKCDGC3R2" received FIN.
[22:34:47 DBG] Connection id "0HMMOKCDGC3R2" disconnecting.
[22:34:47 DBG] Connection id "0HMMOKCDGC3R2" sending FIN because: "The client closed the connection."
```

Feel free to ask questions if I'm not being explicit enough. any suggestions welcome. I have access to modify any layer of this application, but need to keep AAG in place, and need all my other services hosted on aks. 

Thanks!

**Comment by MihaZupan** [comment]
Is the traffic between AAG and YARP HTTP/1.1 or HTTP/2?

From the logs, do these 502s correlate to AAG closing the connection to YARP before the request is received? Or do you see the requests, but the connection is closed before you are able to send a response?

Are these 502s visible to the clients connecting to AAG, or just through AAG's logs?
That is, if you try and make requests to your service, do you see such 502s?

> I don't think this is technically a fix, but I considered reducing YARP's keep alive timeouts so that it closes connections before AAG. I'm not sure if its a placebo, but doing this has seemed to reduce the number of occurrences of 502.

It's entirely possible that you are hitting against the Keep-Alive race condition.
Generally the problem there is that the server's keep-alive is **shorter** than the client's, leading to in-flight requests getting lost. In such cases, it's the server closing the connection on the client, not the other way around.
As such, I would expect that lowering the `KeepAliveTimeout` on the server would lead to eveb more such failures.

As a test, can you try increasing the `KeepAliveTimeout` instead?

**Comment by nquandt** [comment]
The traffic is HTTP/1.1 between AAG and YARP. I (as an end user/client) connecting to AAG would see 502 presented by AAG.

I see the request in YARP and the conn is closed before yarp is able to respond. You are correct I was backwards with the timeouts.. I have increased the Kestrel KeepAliveTimeout to 90 seconds, and that seems to have fixed the issue. I now rarely experience 502. every now and then I see a 504. 

I believe that Azure closes connection pools after 75 seconds, hence why I picked 90.
